File: hooks/python/dere-stop-hook.py
#!/usr/bin/env python3
import json
import os
import sys

# Add the hooks directory to Python path
hooks_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, hooks_dir)

try:
    from rpc_client import RPCClient
except ImportError:
    # Try different path locations
    import sys

    possible_paths = [
        os.path.join(hooks_dir, "..", "..", "hooks", "python"),
        "/home/justin/.local/bin",
        "/home/justin/.config/dere/.claude/hooks",
    ]
    for path in possible_paths:
        if os.path.exists(os.path.join(path, "rpc_client.py")):
            sys.path.insert(0, path)
            from rpc_client import RPCClient

            break
    else:
        raise ImportError("Could not find rpc_client module")


def read_transcript(transcript_path):
    """Read and parse the Claude Code transcript file (JSONL format)"""
    try:
        with open(transcript_path) as f:
            lines = f.readlines()

        # Parse each line as JSON
        transcript_entries = []
        for line in lines:
            line = line.strip()
            if line:
                try:
                    entry = json.loads(line)
                    transcript_entries.append(entry)
                except json.JSONDecodeError:
                    continue

        return transcript_entries
    except Exception as e:
        with open("/tmp/dere_stop_hook_debug.log", "a") as f:
            f.write(f"Error reading transcript: {e}\n")
        return []


def extract_claude_response(transcript_entries):
    """Extract Claude's latest response from transcript entries"""
    try:
        # Look for the latest assistant message
        for entry in reversed(transcript_entries):
            if entry.get("type") == "assistant" and "message" in entry:
                message = entry["message"]
                if message.get("role") == "assistant" and "content" in message:
                    content = message["content"]
                    if isinstance(content, list) and len(content) > 0:
                        # Extract text from content array
                        text_parts = []
                        for item in content:
                            if isinstance(item, dict) and item.get("type") == "text":
                                text_parts.append(item.get("text", ""))
                        return "\n".join(text_parts) if text_parts else None
                    elif isinstance(content, str):
                        return content
        return None
    except Exception as e:
        with open("/tmp/dere_stop_hook_debug.log", "a") as f:
            f.write(f"Error extracting Claude response: {e}\n")
        return None


def main():
    # Debug: log all input
    with open("/tmp/dere_stop_hook_debug.log", "a") as f:
        f.write(f"Stop hook called with args: {sys.argv}\n")

    # Read from stdin (Claude Code passes hook data via stdin)
    try:
        stdin_data = sys.stdin.read().strip()
        with open("/tmp/dere_stop_hook_debug.log", "a") as f:
            f.write(f"Stop hook stdin data: {stdin_data}\n")

        if not stdin_data:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("No stdin data received\n")
            sys.exit(0)

        hook_data = json.loads(stdin_data)

        # Only process if DERE_PERSONALITY is set (i.e., this is a dere session)
        personality = os.getenv("DERE_PERSONALITY")
        if not personality:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("Skipping - not a dere session (no DERE_PERSONALITY)\n")
            sys.exit(0)

        # Get session ID from environment
        session_id = int(os.getenv("DERE_SESSION_ID", "0"))
        project_path = os.getenv("PWD", "")

        # Extract transcript path
        transcript_path = hook_data.get("transcript_path")
        if not transcript_path:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("No transcript path provided\n")
            sys.exit(0)

        # Read and parse transcript
        transcript_entries = read_transcript(transcript_path)
        if not transcript_entries:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("No transcript entries found\n")
            sys.exit(0)

        # Extract Claude's response
        claude_response = extract_claude_response(transcript_entries)
        if not claude_response:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("No Claude response found in transcript\n")
            sys.exit(0)

        with open("/tmp/dere_stop_hook_debug.log", "a") as f:
            f.write(f"Captured Claude response (length: {len(claude_response)})\n")

        # Store Claude's response
        rpc = RPCClient()
        result = rpc.capture_claude_response(session_id, personality, project_path, claude_response)

        with open("/tmp/dere_stop_hook_debug.log", "a") as f:
            f.write(f"RPC result for Claude response: {result}\n")

        if result:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("Claude response captured successfully\n")
        else:
            with open("/tmp/dere_stop_hook_debug.log", "a") as f:
                f.write("Failed to capture Claude response\n")

        # Suppress output to avoid cluttering message history
        print(json.dumps({"suppressOutput": True}))

    except Exception as e:
        with open("/tmp/dere_stop_hook_debug.log", "a") as f:
            f.write(f"Error in stop hook: {e}\n")
        # Suppress output even on error
        print(json.dumps({"suppressOutput": True}))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: plugins/dere_core/hooks/dere-stop-hook.ts
import { appendFileSync } from "node:fs";
import { readFile } from "node:fs/promises";

import { RPCClient } from "./rpc_client.js";

type TranscriptEntry = {
  type?: string;
  message?: {
    role?: string;
    content?: string | Array<{ type?: string; text?: string }>;
  };
};

function logDebug(message: string): void {
  try {
    appendFileSync("/tmp/dere_stop_hook_debug.log", `${message}\n`);
  } catch {
    // ignore logging failures
  }
}

async function readTranscript(transcriptPath: string): Promise<TranscriptEntry[]> {
  try {
    const data = await readFile(transcriptPath, "utf-8");
    const lines = data.split("\n");
    const entries: TranscriptEntry[] = [];

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }
      try {
        entries.push(JSON.parse(trimmed) as TranscriptEntry);
      } catch {
        continue;
      }
    }

    return entries;
  } catch (error) {
    logDebug(`Error reading transcript: ${String(error)}`);
    return [];
  }
}

function extractClaudeResponse(entries: TranscriptEntry[]): string | null {
  try {
    for (let index = entries.length - 1; index >= 0; index -= 1) {
      const entry = entries[index];
      if (entry?.type !== "assistant") {
        continue;
      }
      const message = entry.message;
      if (!message || message.role !== "assistant" || message.content == null) {
        continue;
      }

      const content = message.content;
      if (typeof content === "string") {
        return content;
      }
      if (Array.isArray(content)) {
        const textParts = content
          .filter((item) => item?.type === "text")
          .map((item) => item.text ?? "")
          .filter(Boolean);
        return textParts.length ? textParts.join("\n") : null;
      }
    }
    return null;
  } catch (error) {
    logDebug(`Error extracting Claude response: ${String(error)}`);
    return null;
  }
}

async function main(): Promise<void> {
  logDebug(`Stop hook called with args: ${process.argv.join(" ")}`);

  try {
    const stdin = (await Bun.stdin.text()).trim();
    logDebug(`Stop hook stdin data: ${stdin}`);

    if (!stdin) {
      logDebug("No stdin data received");
      return;
    }

    const hookData = JSON.parse(stdin) as Record<string, unknown>;

    const personality = process.env.DERE_PERSONALITY;
    if (!personality) {
      logDebug("Skipping - not a dere session (no DERE_PERSONALITY)");
      return;
    }

    const sessionId = Number.parseInt(process.env.DERE_SESSION_ID ?? "0", 10);
    const projectPath = process.env.PWD ?? "";
    const transcriptPath =
      typeof hookData.transcript_path === "string" ? hookData.transcript_path : "";

    if (!transcriptPath) {
      logDebug("No transcript path provided");
      return;
    }

    const transcriptEntries = await readTranscript(transcriptPath);
    if (!transcriptEntries.length) {
      logDebug("No transcript entries found");
      return;
    }

    const claudeResponse = extractClaudeResponse(transcriptEntries);
    if (!claudeResponse) {
      logDebug("No Claude response found in transcript");
      return;
    }

    logDebug(`Captured Claude response (length: ${claudeResponse.length})`);

    const rpc = new RPCClient();
    const result = await rpc.captureClaudeResponse(
      sessionId,
      personality,
      projectPath,
      claudeResponse,
    );

    logDebug(`RPC result for Claude response: ${JSON.stringify(result)}`);

    if (result) {
      logDebug("Claude response captured successfully");
    } else {
      logDebug("Failed to capture Claude response");
    }

    console.log(JSON.stringify({ suppressOutput: true }));
  } catch (error) {
    logDebug(`Error in stop hook: ${String(error)}`);
    console.log(JSON.stringify({ suppressOutput: true }));
    process.exit(1);
  }
}

if (import.meta.main) {
  void main();
}
